allennlp/training/callback_trainer.py

```diff
@@ -0,0 +1,250 @@
+        dataset and enable early stopping. There are many other bells and whistles as well.
+
+        Parameters
+        ----------
+        model : ``Model``, required.
+            An AllenNLP model to be optimized. Pytorch Modules can also be optimized if
+            their ``forward`` method returns a dictionary with a "loss" key, containing a
+            scalar tensor representing the loss function to be optimized.
+
+            If you are training your model using GPUs, your model should already be
+            on the correct device. (If you use `Trainer.from_params` this will be
+            handled for you.)
+        optimizer : ``torch.nn.Optimizer``, required.
+            An instance of a Pytorch Optimizer, instantiated with the parameters of the
+            model to be optimized.
+        num_epochs : int, optional (default = 20)
+            Number of training epochs.
+        serialization_dir : str, optional (default=None)
+            Path to directory for saving and loading model files. Models will not be saved if
+            this parameter is not passed.
+        model_save_interval : ``float``, optional (default=None)
+            If provided, then serialize models every ``model_save_interval``
+            seconds within single epochs.  In all cases, models are also saved
+            at the end of every epoch if ``serialization_dir`` is provided.
+        cuda_device : ``Union[int, List[int]]``, optional (default = -1)
+            An integer or list of integers specifying the CUDA device(s) to use. If -1, the CPU is used.
+        """
+        super().__init__(serialization_dir, cuda_device)
+
+        # This is all state that the callbacks might want:
+        # I am not calling move_to_gpu here, because if the model is
+        # not already on the GPU then the optimizer is going to be wrong.
+        self.model = model
+        self.optimizer = optimizer
+        self.validate = False
+
+        # For capturing mid / end-of-epoch metrics
+        self.train_metrics: Dict[str, float] = {}
+        self.val_metrics: Dict[str, float] = {}
+        self.latest_val_metric = 0.0
+        self.train_loss = 0.0
+
+        # For capturing overall metrics
+        self.metrics: Dict[str, Any] = {}
+
+        self.batch_num_total = 0
+        self.batch_group: List[TensorDict] = []
+        self.batches_this_epoch = 0
+
+        self.training_batches: Iterable[List[TensorDict]] = ()
+        self.num_training_batches = 0
+
+        self.should_stop_early = False
+        self.num_epochs = num_epochs
+
+        self.training_start_time = 0.0
+        self.checkpoint_epoch: Union[int, str] = 0
+        self.model_save_interval = model_save_interval
+
+        self.last_log = 0.0
+        self.epoch_number = 0
+        self.batch_grad_norm: Optional[float] = None
+        self.handler = CallbackHandler(callbacks, self)
+
+    def batch_loss(self, batch_group: List[TensorDict], for_training: bool) -> torch.Tensor:
```

Comment:
Why does this function exist here?  It's not called anywhere internally to this class.  Shouldn't it live in a callback somewhere?

Reply:
it lives here so that both the "training" and "validation" callbacks (as well as anyone else who wants it, I guess) can use it

Reply:
Slight preference for having it be a utility function, so that the `CallbackTrainer` can more cleanly get away with `self.model = None`.